File: ui/server/src/services/worker-scanner.ts
import { readdir, readFile, stat } from "fs/promises";
import { join } from "path";
import { execSync } from "child_process";

export interface Worker {
  id: string;
  taskId: string;
  timestamp: number;
  pid: number | null;
  status: "running" | "stopped" | "completed" | "failed";
  prdPath: string;
  logPath: string;
  workspacePath: string;
  prUrl: string | null;
}

export interface OrchestratorStatus {
  running: boolean;
  pid: number | null;
}

function isProcessRunning(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}

function isWorkerProcess(pid: number): boolean {
  try {
    const result = execSync(`ps -p ${pid} -o args=`, { encoding: "utf-8" });
    return result.includes("lib/worker.sh");
  } catch {
    return false;
  }
}

function isOrchestratorProcess(pid: number): boolean {
  try {
    const result = execSync(`ps -p ${pid} -o args=`, { encoding: "utf-8" });
    return result.includes("wiggum-run");
  } catch {
    return false;
  }
}

async function getPrdStatus(
  prdPath: string
): Promise<"complete" | "failed" | "incomplete"> {
  try {
    const content = await readFile(prdPath, "utf-8");
    if (content.includes("- [*]")) return "failed";
    if (content.includes("- [ ]")) return "incomplete";
    return "complete";
  } catch {
    return "incomplete";
  }
}

export async function scanWorkers(ralphDir: string): Promise<Worker[]> {
  const workersDir = join(ralphDir, "workers");
  const workers: Worker[] = [];

  try {
    const entries = await readdir(workersDir);

    for (const entry of entries) {
      if (!entry.startsWith("worker-")) continue;

      const workerDir = join(workersDir, entry);
      const stats = await stat(workerDir);
      if (!stats.isDirectory()) continue;

      // Parse worker ID: worker-TASK-XXX-TIMESTAMP
      const match = entry.match(/^worker-([A-Za-z]{2,8}-\d+)-(\d+)$/);
      if (!match) continue;

      const [, taskId, timestampStr] = match;
      const timestamp = parseInt(timestampStr, 10);

      const pidPath = join(workerDir, "worker.pid");
      const prdPath = join(workerDir, "prd.md");
      const logPath = join(workerDir, "worker.log");
      const workspacePath = join(workerDir, "workspace");
      const prUrlPath = join(workerDir, "pr_url.txt");

      let pid: number | null = null;
      let status: Worker["status"] = "stopped";

      // Check PID file
      try {
        const pidContent = await readFile(pidPath, "utf-8");
        const parsedPid = parseInt(pidContent.trim(), 10);
        if (!isNaN(parsedPid)) {
          pid = parsedPid;
          if (isProcessRunning(parsedPid) && isWorkerProcess(parsedPid)) {
            status = "running";
          }
        }
      } catch {
        // No PID file
      }

      // If not running, check PRD status
      if (status !== "running") {
        const prdStatus = await getPrdStatus(prdPath);
        if (prdStatus === "complete") {
          status = "completed";
        } else if (prdStatus === "failed") {
          status = "failed";
        }
      }

      // Check for PR URL
      let prUrl: string | null = null;
      try {
        prUrl = (await readFile(prUrlPath, "utf-8")).trim();
      } catch {
        // No PR URL file
      }

      workers.push({
        id: entry,
        taskId,
        timestamp,
        pid,
        status,
        prdPath,
        logPath,
        workspacePath,
        prUrl,
      });
    }
  } catch {
    // Workers directory doesn't exist
  }

  // Sort by timestamp descending (newest first)
  workers.sort((a, b) => b.timestamp - a.timestamp);

  return workers;
}

export async function getOrchestratorStatus(
  ralphDir: string
): Promise<OrchestratorStatus> {
  const pidPath = join(ralphDir, ".orchestrator.pid");

  try {
    const pidContent = await readFile(pidPath, "utf-8");
    const pid = parseInt(pidContent.trim(), 10);

    if (!isNaN(pid) && isProcessRunning(pid) && isOrchestratorProcess(pid)) {
      return { running: true, pid };
    }
  } catch {
    // No PID file
  }

  return { running: false, pid: null };
}

File: tui/wiggum_tui/data/worker_scanner.py
"""Worker scanner - ported from TypeScript."""

import os
import re
import subprocess
from pathlib import Path
from .models import Worker, WorkerStatus


# Pattern: worker-TASK-XXX-TIMESTAMP
WORKER_PATTERN = re.compile(r"^worker-([A-Za-z]{2,10}-\d{1,4})-(\d+)$")


def is_process_running(pid: int) -> bool:
    """Check if a process is running."""
    try:
        os.kill(pid, 0)
        return True
    except ProcessLookupError:
        return False
    except PermissionError:
        return True


def batch_check_processes(pids: list[int]) -> dict[int, str]:
    """Check multiple PIDs in a single ps call.

    Args:
        pids: List of process IDs to check.

    Returns:
        Dictionary mapping running PIDs to their command args.
        PIDs not in the dict are not running.
    """
    if not pids:
        return {}

    try:
        # Single ps call for all PIDs - much faster than per-PID calls
        result = subprocess.run(
            ["ps", "-p", ",".join(str(p) for p in pids), "-o", "pid=,args="],
            capture_output=True,
            text=True,
            timeout=5,
        )
        process_info: dict[int, str] = {}
        for line in result.stdout.strip().split("\n"):
            if not line.strip():
                continue
            parts = line.strip().split(None, 1)
            if len(parts) >= 2:
                try:
                    pid = int(parts[0])
                    args = parts[1]
                    process_info[pid] = args
                except ValueError:
                    continue
        return process_info
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return {}


def is_worker_process(pid: int, process_info: dict[int, str] | None = None) -> bool:
    """Check if PID is actually a worker process.

    Args:
        pid: Process ID to check.
        process_info: Optional pre-fetched process info from batch_check_processes().
            If not provided, will make a subprocess call (slower).
    """
    if process_info is not None:
        args = process_info.get(pid, "")
        return "bash" in args

    # Fallback to single-PID check
    try:
        result = subprocess.run(
            ["ps", "-p", str(pid), "-o", "args="],
            capture_output=True,
            text=True,
            timeout=5,
        )
        # Agents run in bash subshells via run_agent()
        return "bash" in result.stdout
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return False


def is_orchestrator_process(pid: int, process_info: dict[int, str] | None = None) -> bool:
    """Check if PID is actually an orchestrator process.

    Args:
        pid: Process ID to check.
        process_info: Optional pre-fetched process info from batch_check_processes().
            If not provided, will make a subprocess call (slower).
    """
    if process_info is not None:
        args = process_info.get(pid, "")
        return "wiggum-run" in args

    # Fallback to single-PID check
    try:
        result = subprocess.run(
            ["ps", "-p", str(pid), "-o", "args="],
            capture_output=True,
            text=True,
            timeout=5,
        )
        return "wiggum-run" in result.stdout
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return False


def get_prd_status(prd_path: Path) -> str:
    """Get status from PRD file content.

    Returns: 'complete', 'failed', or 'incomplete'
    """
    try:
        content = prd_path.read_text()
        if "- [*]" in content:
            return "failed"
        if "- [ ]" in content:
            return "incomplete"
        return "complete"
    except (OSError, FileNotFoundError):
        return "incomplete"


def scan_workers(ralph_dir: Path) -> list[Worker]:
    """Scan .ralph/workers directory for workers.

    Args:
        ralph_dir: Path to .ralph directory.

    Returns:
        List of Worker objects, sorted by timestamp (newest first).
    """
    workers_dir = ralph_dir / "workers"
    workers: list[Worker] = []

    if not workers_dir.is_dir():
        return workers

    # First pass: collect all worker directories and their PIDs
    worker_entries: list[tuple[Path, str, int, int | None]] = []  # (dir, task_id, timestamp, pid)
    all_pids: list[int] = []

    for entry in workers_dir.iterdir():
        if not entry.name.startswith("worker-"):
            continue
        if not entry.is_dir():
            continue

        # Parse worker ID
        match = WORKER_PATTERN.match(entry.name)
        if not match:
            continue

        task_id, ts_str = match.groups()
        # Use timestamp from directory name (creation time) for accurate start time
        try:
            timestamp = int(ts_str)
        except ValueError:
            timestamp = 0

        pid_path = entry / "agent.pid"
        pid: int | None = None

        if pid_path.exists():
            try:
                pid_content = pid_path.read_text().strip()
                pid = int(pid_content)
                all_pids.append(pid)
            except (ValueError, OSError):
                pass

        worker_entries.append((entry, task_id, timestamp, pid))

    # Batch check all PIDs in a single subprocess call
    process_info = batch_check_processes(all_pids)

    # Second pass: build Worker objects using batch results
    for entry, task_id, timestamp, pid in worker_entries:
        prd_path = entry / "prd.md"
        log_path = entry / "worker.log"
        workspace_path = entry / "workspace"
        pr_url_path = entry / "pr_url.txt"

        status = WorkerStatus.STOPPED

        # Check if process is running using batch results
        if pid is not None and is_worker_process(pid, process_info):
            status = WorkerStatus.RUNNING

        # If not running, check PRD status
        if status != WorkerStatus.RUNNING:
            prd_status = get_prd_status(prd_path)
            if prd_status == "complete":
                status = WorkerStatus.COMPLETED
            elif prd_status == "failed":
                status = WorkerStatus.FAILED

        # Check for PR URL
        pr_url: str | None = None
        if pr_url_path.exists():
            try:
                pr_url = pr_url_path.read_text().strip()
                if not pr_url:
                    pr_url = None
            except OSError:
                pass

        workers.append(
            Worker(
                id=entry.name,
                task_id=task_id,
                timestamp=timestamp,
                pid=pid,
                status=status,
                prd_path=str(prd_path),
                log_path=str(log_path),
                workspace_path=str(workspace_path),
                pr_url=pr_url,
            )
        )

    # Sort by timestamp descending (newest first)
    workers.sort(key=lambda w: w.timestamp, reverse=True)

    return workers


def get_orchestrator_status(ralph_dir: Path) -> tuple[bool, int | None]:
    """Get orchestrator status.

    Args:
        ralph_dir: Path to .ralph directory.

    Returns:
        Tuple of (running: bool, pid: int | None).
    """
    pid_path = ralph_dir / ".orchestrator.pid"

    if not pid_path.exists():
        return False, None

    try:
        pid_content = pid_path.read_text().strip()
        pid = int(pid_content)

        if is_process_running(pid) and is_orchestrator_process(pid):
            return True, pid
    except (ValueError, OSError):
        pass

    return False, None


def get_worker_counts(workers: list[Worker]) -> dict[str, int]:
    """Get count of workers in each status.

    Args:
        workers: List of workers.

    Returns:
        Dictionary with counts for each status.
    """
    counts = {
        "running": 0,
        "stopped": 0,
        "completed": 0,
        "failed": 0,
        "total": len(workers),
    }
    for worker in workers:
        counts[worker.status.value] += 1
    return counts


def get_running_worker_for_task(ralph_dir: Path, task_id: str) -> Worker | None:
    """Find a running worker for a specific task ID.

    Args:
        ralph_dir: Path to .ralph directory.
        task_id: Task ID to search for (e.g., "TASK-001").

    Returns:
        Worker if a running worker is found for the task, None otherwise.
    """
    workers = scan_workers(ralph_dir)
    for worker in workers:
        if worker.task_id == task_id and worker.status == WorkerStatus.RUNNING:
            return worker
    return None


def get_task_running_status(ralph_dir: Path, task_ids: list[str]) -> dict[str, tuple[bool, int | None]]:
    """Get running status for multiple tasks at once (more efficient).

    Args:
        ralph_dir: Path to .ralph directory.
        task_ids: List of task IDs to check.

    Returns:
        Dictionary mapping task_id to (is_running, start_timestamp).
        start_timestamp is the worker directory mtime when running.
    """
    workers = scan_workers(ralph_dir)
    result: dict[str, tuple[bool, int | None]] = {}

    # Build a map of task_id -> running worker
    running_workers: dict[str, Worker] = {}
    for worker in workers:
        if worker.status == WorkerStatus.RUNNING:
            # Use the most recent worker for each task
            if worker.task_id not in running_workers:
                running_workers[worker.task_id] = worker

    for task_id in task_ids:
        if task_id in running_workers:
            worker = running_workers[task_id]
            result[task_id] = (True, worker.timestamp)
        else:
            result[task_id] = (False, None)

    return result
